src/auth/src/credentials/external_account.rs

```diff
@@ -0,0 +1,398 @@
+/// #     "test-only".to_string()
+/// # }
+/// # fn workload_identity_pool() -> String {
+/// #     "test-only".to_string()
+/// # }
+/// # fn workload_identity_provider() -> String {
+/// #     "test-only".to_string()
+/// # }
+/// ```
+pub struct Builder {
+    external_account_config: Value,
+    quota_project_id: Option<String>,
+    scopes: Option<Vec<String>>,
+}
+
+impl Builder {
+    /// Creates a new builder using [external_account_credentials] JSON value.
+    ///
+    /// [external_account_credentials]: https://google.aip.dev/auth/4117#configuration-file-generation-and-usage
+    pub fn new(external_account_config: Value) -> Self {
+        Self {
+            external_account_config,
+            quota_project_id: None,
+            scopes: None,
+        }
+    }
+
+    /// Sets the [quota project] for this credentials.
+    ///
+    /// In some services, you can use a service account in
+    /// one project for authentication and authorization, and charge
+    /// the usage to a different project. This requires that the
+    /// service account has `serviceusage.services.use` permissions on the quota project.
+    ///
+    /// [quota project]: https://cloud.google.com/docs/quotas/quota-project
+    pub fn with_quota_project_id<S: Into<String>>(mut self, quota_project_id: S) -> Self {
+        self.quota_project_id = Some(quota_project_id.into());
+        self
+    }
+
+    /// Overrides the [scopes] for this credentials.
+    ///
+    /// [scopes]: https://developers.google.com/identity/protocols/oauth2/scopes
+    pub fn with_scopes<I, S>(mut self, scopes: I) -> Self
+    where
+        I: IntoIterator<Item = S>,
+        S: Into<String>,
+    {
+        self.scopes = Some(scopes.into_iter().map(|s| s.into()).collect());
+        self
+    }
+
+    /// Returns a [Credentials] instance with the configured settings.
+    ///
+    /// # Errors
+    ///
+    /// Returns a [CredentialsError] if the `external_account_config`
+    /// provided to [`Builder::new`] cannot be successfully deserialized into the
+    /// expected format for an external account configuration. This typically happens if the
+    /// JSON value is malformed or missing required fields. For more information,
+    /// on the expected format, consult the relevant section in the
+    /// [external_account_credentials] guide.
+    ///
+    /// [external_account_credentials]: https://google.aip.dev/auth/4117#configuration-file-generation-and-usage
+    pub fn build(self) -> BuildResult<Credentials> {
+        let external_account_config: ExternalAccountConfig =
+            serde_json::from_value(self.external_account_config).map_err(BuilderError::parsing)?;
+
+        let mut config = external_account_config.clone();
+        if let Some(scopes) = self.scopes {
+            config.scopes = Some(scopes);
+        }
+
+        let token_provider = ExternalAccountTokenProvider {
+            subject_token_provider: external_account_config.credential_source,
```

Comment:
Is it possible to identify the correct byoid cred type and assign the appropriate subject token provider here? It seems unnecessary to go through the process of identifying each time subject token is needed.